### PR TITLE
Decoder refactor + fixing equations

### PR DESCRIPTION
**Pull request by jindrahelcl**
- added bias term in attention 
- refactored decoder to have configurable output projection (see decoders.output_projection module)

See also the commit messages for a little bit more info

@jlibovicky Please check the multilayer perceptron implementations and tell me if I can remove the old one. (And possibly, where to update the older code that was using the MLP class)

### CODE REVIEW

**Comment by jlibovicky** [comment]
Tady jsou taky gradienty NaN už od commitu e90c8d8, ale vypadá to docela nedeterministicky - padá po různé době učení a dokonce jsem zažil, že to nespadlo. Zkoušel jsem, jestli si to nenese tuhle vlastnost už z materu, ale vypadá to, že ne (tam jsem seběhnul testy 10× po sobě) bez problémů.

**Comment by jindrahelcl** [comment]
To vypadá na nějakej jinej problém, tady v tom případě musí bejt len(outputs) == 0 a to způsobí ten nan... Thle píše travis log:

```
/home/travis/build/ufal/neuralmonkey/bin/neuralmonkey/runners/base_runner.py:59: RuntimeWarning: divide by zero encountered in double_scalars
  losses = [l / len(outputs) for l in losses_sum]
```

**Comment by jlibovicky** [comment]
Rebasoval jsem je. Je to kód v base_runneru, protože je tam funkce, které umí agregovat ExecutionResulty dohromady.

Tohle se děje v numpy, když se zavolá trainer, kterej ale nedává outputy a nepočítá losy, takže se tam dělí nuly nulou a hodí to warning, ale nikde dál se to nepoužije. To prostě končí tím, že je tam ExecutionResult, ten má v sobě NaNy a to je konec story. (Ale samozřejmě to můžu opravit :-D)

Tyhlety NaNy musí vznikat vevnitř v TensorFlow, jinak by se nemohly objevit v těch summaries.

**Comment by jlibovicky** [comment]
Navíc se tenhleten warning píše i v masteru a tam mi to nepadá, ani když to pouštím pořád dokola.

**Comment by jlibovicky** [comment]
Děje se to jenom s `bahdanau.ini`, se `small.ini` to nepadá.

**Comment by jlibovicky** [comment]
A tak dlouho jsem diffoval ty konfiguráky, až si myslím, že je to těmi miniaturními velikostmi stavů.

**Comment by jindrahelcl** [comment]
Fakt, jo? Tak je zkus zvětšit, no.. ale je to divný.. Možná to učení vyexploduje než se to stačí ustálit..

**Comment by jlibovicky** [comment]
Tak ani zvětšení těch stavů nepomohlo. Ale teď už mi to fakt připadá, že jedinej diff jsou proházený řádky.

**Comment by jlibovicky** [comment]
Zkus se v tom vrtat chvíli ty :-P

**Comment by jindrahelcl** [comment]
Já se v tom vrtám.. :-)

**Comment by jindrahelcl** [comment]
No každopádně to před tim rebasem, když to bylo na masteru, tak to tohle nedělalo.. (Ale stejně to nefunguje, takže vlsatně vůbec teď nevim, kam se koukat)

**Comment by jindrahelcl** [comment]
Proč mi to teď v masteru funguje??

**Comment by jindrahelcl** [comment]
Vypadá to, že tam dělá bordel ten bias na attention..

**Comment by jlibovicky** [comment]
Větev s těma variables jsem spravil (bylo to nulovou inicializací vektoru, s jehož pomocí se pak počítala podobnost vektorů), to jsem mergnul a tuhle větev jsem rebasoval s masterem. Zkusil jsem dát taky do háje ten bias (o kterým si myslím, že tam nemůže dělat nic užitečnýho) a nic se nezměnilo.

**Comment by jindrahelcl** [comment]
ok.. Koukni prosím do bahdanau článku  na stránku 14 dole.. Je tam odstavec o inicializaci a věta začínající "All the elements of V_a".. V_a ve článku nikde neni, je tam jen v_a - to je ten vektor, o kterym mluvíš. A tam píšou, že to inicializujou nulama. Tak nevim

**Comment by jindrahelcl** [comment]
Je možný, že tim mysleli buď v_a nebo V_o...

**Comment by jindrahelcl** [comment]
Ad bias - ten jsem tam přidal, protože jsem ho našel v nematusovi. Asi na něm moc nezáleží, ale nechal bych ho tam

**Comment by jlibovicky** [comment]
Tak si rozmysli, co se stane, když ho inicializuješ nulama: vektor, se kterým potom měříš podobnost stavu dekodéru bude vždycky nula, takže ti tam nikdy nedoteče žádnej gradient při bakckpropagaci, derivaci není možný spočítat - proto je NaN. To potom spadne, když se histogram gradients pokusí ten NaN nakreslit. Výpočet jako takovej to zdánlivě neovlivní, protože ty váhy, co se vynanujou jsou odstřihlý těma nulama od zbyteku grafu a nikdy se ve výpočtu neprojeví. V týhle větvi si myslím, že to bude nějaká podobná záležitost.

**Comment by jindrahelcl** [comment]
To dává smysl. Tim pádem ale nevim, co myslej tim V_a, protože i tim se něco násobí (embeddingy), takže to by se taky nic neučilo, ne?

Jinak ale moc nerozumim tomu, proč když je parametr nulovej, proč do něj nedoteče gradient - biasy se taky inicializujou nulama, a gradienty do nich jdou.

**Comment by jindrahelcl** [comment]
Nebo nesmí vzniknout nula jenom z něčeho, do čeho leze (nějak transformovanej) vstup?

**Comment by jlibovicky** [comment]
Ten důvod je jednoduchej, biasy se přičítaj, těmahle nulama se násobilo.

**Comment by jindrahelcl** [comment]
No jó. Napsal jsem si to na papírek a dávám ti za pravdu.. Chápu to teď správně, že žádná váha, kterou se násobí, by neměla bejt nula, nebo stačí, že na tom řádku v matici nebudou samý nuly? Taky by to totiž mohlo bejt tim, že se do tý inicializace dávaj docela malý rozptyly, takže tam může těch nul hodně vzniknout.

**Comment by jlibovicky** [comment]
Ty, co se nainicializujou nulový, zůstanou nulový napořád, ale když jich je pár, tak to nevadí. (A když se to stane v průběhu učení, tak to Adamové nějak vyřeší, protože je tam nějaká setrvačnost.) Přišel jsem ještě na to, že TF by se poradil s tím, že se to vynuluje (gradienty by prostě byly nula), ale to místo, kde se to doopravdy vynanovalo, je gradient clipping, protože tam se dělilo nulovou normou.

**Comment by jindrahelcl** [comment]
Hmm, tak ty v tom svym generic traineru dáváš clip_norm defaultně na False a pak clipuješ když clip_norm is not None. :-)

**Comment by jindrahelcl** [comment]
tests/small.ini fungovalo, protože je tam clip_norm nastavený na 1.0, ale bahdanau ho neměl, takže se použilo False

**Comment by jindrahelcl** [comment]
Ale teď jsem přišel na papírku na protipříklad:

Máš model: y = p * x + 20
Data: (x, y_gold) = (3, 26)
Loss funkci (MSE) L = 1/2 (y - y_gold) ** 2

Derivace se spočtou takle:
dL/dy = (y - y_gold)
dy/dp = x
dL/dp = dL/dy * dy/dp = (y - y_gold) * x

Když p nastavíš na 0, první iterace bude vypadat takle:
L = -6
dL/dp = 3 * (-6) = -18

a krok učení s alfa=0.1 by byl p = p + 1.8

Takže nemam nulovej gradient, i když sem to tim péčkem přenásoboval.

**Comment by jindrahelcl** [comment]
A gradient clipping nany nevrací, když je L2 nulová - jediný, kde tam můžou vzniknout nany je když je clip_norm nulová.

**Comment by jlibovicky** [comment]
Ty nuly si musím ještě pořádně rozmyslet, ale asi máš pravdu. Takže jediná chyba bylo, že se tam klipovaly gradienty na False, co se převedlo na nulu?

File: neuralmonkey/trainers/generic_trainer.py [hunk 1]
```diff
@@ -26,10 +26,7 @@ def __init__(self, objectives: List[Objective],
                  l1_weight=0.0, l2_weight=0.0,
                  clip_norm=False, optimizer=None) -> None:
 
-        if optimizer is None:
-            self.optimizer = tf.train.AdamOptimizer(1e-4)
-        else:
-            self.optimizer = optimizer
+        self.optimizer = optimizer or tf.train.AdamOptimizer(1e-4)
```
Comment (jindrahelcl) [inline neuralmonkey/trainers/generic_trainer.py]: To je jen refactor (doufám!!)

Comment (jindrahelcl) [inline neuralmonkey/trainers/generic_trainer.py]: Když optimizer is None tak se tam dá Adam, jinak se použije

**Comment by jindrahelcl** [comment]
Myslim že jo. I jsem to vyzkoušel v masteru vrátit tam tu nulovou inicializaci a test prošel..

File: neuralmonkey/trainers/generic_trainer.py (hunk 1 above)
Comment (jlibovicky) [inline neuralmonkey/trainers/generic_trainer.py]: Jo, nevšim jsem si toho oru. Cool.

**Comment by jindrahelcl** [comment]
Jestli s tim teda jsi spokojenej, tak to prosim mergni, ja uz do vecera zas nebudu na netu..
